### PR TITLE
signin/signup endpoint

### PR DESCRIPTION
**Pull request by EngrEric**

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -17,10 +22,19 @@ app.use(express.static(__dirname));
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 
+//Cross-Origin Resource Sharing CORS
+//To enable access of resources from our server
+
+if(process.env.NODE_ENV === 'development'){
```
Comment (Way2nnadi) [inline server.js]: this should be more like `if (process.env.NODE_ENV !== 'production')`. Unless you have something somewhere that is setting the `NODE_ENV` to`development`.

File: controllers/signInAndOutController.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            });
+        }
+
+        email = email.toLowerCase();
+        email = email.trim();
+
+        User.find({
+            email,
+            password
+        }, (err, users) => {
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: try to maximize each line
```
User.find({ email, password }), (err, user) => {
...
}
...
```
this is much cleaner, readable and uses less lines

File: controllers/signInAndOutController.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+module.exports = {
+    
+    signInUser: (req, res) => {
+
+        const { body } = req;
+
+        const { password } = body;
+
+        let { email } = body;
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: group variable definitions together, without the extra spaces in between

File: controllers/productsController.js [hunk 4]
```diff
@@ -1,76 +1,76 @@
...
-  productUpdate: async (req, res) => {
-    try {
-      const data = await Product.findOneAndUpdate({ _id: req.params.productsId }, req.body, { new: true });
-      return res.json({ data, success: true, message: 'Product successfully updated' });
-    } catch (err) {
-      res.send({ success: false, err });
+    // updates products using id
+    productUpdate: async (req, res) => {
+        try {
+            const data = await Product.findOneAndUpdate({ _id: req.params.productsId }, req.body, { new: true });
```
Comment (Way2nnadi) [inline controllers/productsController.js]: destructure `req` and simplify the code
```
const { body, params { productsId: _id } } = req;
const data = await Product.findOneAndUpdate({ _id, body, { new: true } });
```

File: controllers/producersController.js [hunk 5]
```diff
@@ -1,85 +1,89 @@
...
-    try {
-      await producer.save();
-      return res.json({ success: true, message: 'Producer successfully added!' });
-    } catch (err) {
-      res.send({ success: false, err });
-    }
-  },
+        try {
+            await producer.save();
+            return res.json({ success: true, message: 'Producer successfully added!' });
```
Comment (Way2nnadi) [inline controllers/producersController.js]: move message string to a `constants.js` file

File: controllers/producersController.js [hunk 6]
```diff
@@ -1,85 +1,89 @@
...
+            const{ body, params:{ producersId: _id } } = req;
+            const data = await Producer.findOneAndUpdate( _id, body, { new: true });
 
-  producerUpdate: async(req, res) => {
-    try {
-      const data = await Producer.findOneAndUpdate({ _id: req.params.producersId }, req.body, { new: true });
-      return res.json({ data, success: true, message: 'Producer successfully updated!' });
-    } catch (err) {
-      res.send({ success: false, err });
+            return res.json({ data, success: true, message: 'Producer successfully updated!' });
```
Comment (Way2nnadi) [inline controllers/producersController.js]: move message string to a `constants.js` file

File: controllers/productsController.js [hunk 7]
```diff
@@ -1,76 +1,80 @@
...
-    });
+        product = Object.assign(product, {
+            name,
+            quantity,
+            producerId,
+            typeOfProduct,
+            createdAt,
+            updatedAt,
+            deletedAt
+        });
```
Comment (Way2nnadi) [inline controllers/productsController.js]:
```
const product = Object.assign(new Product(), {
...
};
```

File: controllers/productsController.js [hunk 8]
```diff
@@ -1,76 +1,80 @@
...
-    try {
-      await product.save();
-      return res.json({ success: true, message: 'Product successfully added!' });
-    } catch (err) {
-      res.send({ success: false, err });
-    }
-  },
+        try {
+            await product.save();
+            return res.json({ success: true, message: 'Product successfully added!' });
```
Comment (Way2nnadi) [inline controllers/productsController.js]: move message string to a `constants.js` file

File: controllers/producersController.js [hunk 9]
```diff
@@ -1,85 +1,89 @@
...
+                phoneNumber,
+                country,
+                state,
+                localGovernment,
+                address,
+                typeOfProducts,
+                createdAt,
+                updatedAt,
+                deletedAt
+            });
```
Comment (Way2nnadi) [inline controllers/producersController.js]:
```
const producer = Object.assign(new Producer, {
...
});
```

File: controllers/productsController.js [hunk 10]
```diff
@@ -1,76 +1,80 @@
...
-    } catch (err) {
-      res.send({ success: false, err });
-    }
-  },
+    // deletes products using id
+    productDelete: async (req, res) => {
+        try {
+            const { body, params: { id: _id } } =  req;
+            await Product.findByIdAndRemove(_id, body);
+            return res.json({ success: true, message: 'Product successfully deleted!' });
```
Comment (Way2nnadi) [inline controllers/productsController.js]: move message string to a `constants.js` file

File: controllers/signInAndOutController.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+        if (!email || !password) {
+            return res.send({
+                success: false,
+                message: 'Error: Email or Password cannot be blank.'
+            });
+        }
+
+        email = email.toLowerCase();
+        email = email.trim();
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: `email = email.toLowerCase().trim();`

File: controllers/signInAndOutController.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+        User.find({ email, password }, (err, users) => {
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]:
```
try {
    const user = await User.findOne({ email, password })
...
} catch( error) {
 ...
}
```

File: controllers/signInAndOutController.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
+const User = require('../models/user');
+const UserSession = require('../models/userSession');
+
+module.exports = {
+    
+    signInUser: (req, res) => {
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: should be an `async/await` function

File: controllers/signInAndOutController.js [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+          
+            const user = users[0];
+            /*   console.log(users);
+          let userLoginDetails = users.map(user=>({email: user.email, password: user.password}));
+
+         if(userLoginDetails[0].email !== email || userLoginDetails[0].password !== password){
+           return res.send({
+             message: 'email or password is incorrect'
+           });
+         } */
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: clean up un-used code

File: controllers/signInAndOutController.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+            // Otherwise correct user
+            const userSession = new UserSession();
+            userSession.userId = user._id;
+
+            userSession.save((err, doc) => {
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: should await for `doc`

File: controllers/signInAndOutController.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    allUsers: async (req, res) => {
+        try {
+            const data = await User.find();
+            return res.json({ success: true, data });
+        } catch (err) {
+            return res.json({ success: false, error: err });
+        }
+    },
+
+    logout: (req, res) => {
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: should be an `async/await` function

File: controllers/signupController.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+            email,
+            username,
+            password
+        } = req.body;
+
+        user = Object.assign(user, {
+            email,
+            username,
+            password
+        });
```
Comment (Way2nnadi) [inline controllers/signupController.js]:
```
const user = Object.assign(new User(), {...})
```

File: controllers/signupController.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+        if(!email || !password){
+            return res.send({ success: false, message: 'email or password cannot be blank' });
+        }
+        /* 
+         email = email.toLowerCase();
+         email = email.trim();
+ */
+        // 1. Verify email doesn't exist
+        User.find({
```
Comment (Way2nnadi) [inline controllers/signupController.js]: use `User.findOne` and make it await for `previousUser`;

File: controllers/signupController.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+            }
+        });
+
+        // Save the new user
+        try {
+            await user.save();
+            return res.send({
+                success: true,
+                message: 'Signed up'
+            });
```
Comment (Way2nnadi) [inline controllers/signupController.js]: `line 45-50` = very good

File: controllers/signupController.js [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+        }catch(err){
```
Comment (Way2nnadi) [inline controllers/signupController.js]: space before `catch`

File: models/user.js [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+
+    password: {
+        type: String,
+    },
+
+},
+
+{ 
+    versionKey:false 
+});
```
Comment (Way2nnadi) [inline models/user.js]: delete extra spacing on line 8, 12, 16, 18

File: server.js [hunk 22]
```diff
@@ -17,10 +22,19 @@ app.use(express.static(__dirname));
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 
+//Cross-Origin Resource Sharing CORS
+//To enable access of resources from our server
```
Comment (Way2nnadi) [inline server.js]: use multiple line commenting style here

File: controllers/producersController.js [hunk 23]
```diff
@@ -1,85 +1,88 @@
...
-      state,
-      localGovernment,
-      address,
-      typeOfProducts,
-      createdAt,
-      updatedAt,
-      deletedAt
-    } = req.body;
+        const producer = Object
+            .assign(new Producer, {
```
Comment (Way2nnadi) [inline controllers/producersController.js]: `new Producer` is not javascript syntax

File: controllers/producersController.js [hunk 24]
```diff
@@ -1,85 +1,88 @@
 const Producer = require('../models/producer');
+const { ADD_PRODUCER, DELETE_PRODUCER, UPDATE_PRODUCER } = require('./constants');
 
 module.exports = {
-  producerCreate: async(req, res) => {
+    producerCreate: async(req, res) => {
     //  post new producer to the database
-    let producer = new Producer();
+        //  body parser lets us use the req.body
```
Comment (Way2nnadi) [inline controllers/producersController.js]: this comment seems out of place

File: controllers/productsController.js [hunk 25]
```diff
@@ -1,76 +1,80 @@
...
-    product = Object.assign(product, {
-      name,
-      quantity,
-      producerId,
-      typeOfProduct,
-      createdAt,
-      updatedAt,
-      deletedAt
-    });
+        const product = Object.assign( new Product, {
```
Comment (Way2nnadi) [inline controllers/productsController.js]: `new Product` is not javascript syntax

File: controllers/productsController.js [hunk 26]
```diff
@@ -1,76 +1,80 @@
...
+            return res.json({ success: true, message: DELETE_PRODUCT });
+        } catch (err) {
+            res.send({ success: false, err });
+        }
+    },
+
+    // updates products using id
+    productUpdate: async (req, res) => {
+        try {
+            /*  const data = await Product.findOneAndUpdate({ _id: req.params.productsId }, req.body, { new: true });
```
Comment (Way2nnadi) [inline controllers/productsController.js]: delete unused code

File: controllers/signInAndOutController.js [hunk 27]
```diff
@@ -0,0 +1,85 @@
+const User = require('../models/user');
+const UserSession = require('../models/userSession');
+
+module.exports = {
+
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: delete extra line

File: controllers/signInAndOutController.js [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+        try {
+            users = await User.findOne({ email, password });
+            if(!users){
+                return res.send({
+                    success: false,
+                    message: 'Cannnot find any user with the email  or password'
+                });
+            } 
+        } catch(err){
+            res.send({ success: false, err });
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: should `return res.send({ success: false, err });`

File: controllers/signInAndOutController.js [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+
+        try{
+            const doc = await userSession.save();
+            return res.send({
+                success: true,
+                message: 'Valid sign in',
+                token: doc._id
+            });
+        } catch(err){
+            res.send({ success: false, err });
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: return here as well

File: controllers/signInAndOutController.js [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+        } catch (err) {
+            return res.json({ success: false, err });
+        }
+    },
+
+    logout: async (req, res) => {
+        // Get the token
+        const { query } = req;
+        const { token } = query;
+        // ?token=test
```
Comment (Way2nnadi) [inline controllers/signInAndOutController.js]: is this comment needed?

File: controllers/signupController.js [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+            password
+        } = req.body;
+        const user = Object.assign(new User(), {
+            email,
+            username,
+            password
+        });
+
+        if(!email || !password){
+            return res.send({ success: false, message: 'email or password cannot be blank' });
```
Comment (Way2nnadi) [inline controllers/signupController.js]: move string to `constants.js`

File: controllers/signupController.js [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        try{
+            const previousUsers = await User.findOne({ email  });
+            if(previousUsers){
+                return res.send({
+                    success: false,
+                    message: 'Error: Account already exist.'
```
Comment (Way2nnadi) [inline controllers/signupController.js]: move string to `constants.js`

File: controllers/signupController.js [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+        } catch(err){
+            res.send({ err });
+        }
+
+        // Save the new user
+        try {
+            await user.save();
+            return res.send({
+                success: true,
+                message: 'Signed up'
```
Comment (Way2nnadi) [inline controllers/signupController.js]: move string to `constants.js`

File: controllers/producersController.js (hunk 23 above)
Comment (EngrEric) [reply controllers/producersController.js]: whats the proper js syntax here.
it should be ```new Producer()```

**Pull request merged**